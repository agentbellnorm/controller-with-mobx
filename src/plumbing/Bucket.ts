import { makeAutoObservable, autorun, runInAction } from 'mobx';

export interface SwapFunction<G> {
    (state: G, ...rest: any[]): G
}

export default class Bucket<T> {
    state!: T;

    constructor(initialState: T) {
        makeAutoObservable(this, {}, { deep: true });
        runInAction(() => {
            this.reset(initialState);
        });
    }

    reset(newValue: T): T {
        this.state = newValue;
        return this.state;
    }

    swap(swapFunction: SwapFunction<T>, ...args: any[]): T {
        const newState = swapFunction.apply(null, [this.state, ...args]);
        return this.reset(newState);
    }

    deref(): T {
        return this.state;
    }

    onChange(listener: (state: T) => void) {
        autorun(() => listener(this.state));
    }
}
