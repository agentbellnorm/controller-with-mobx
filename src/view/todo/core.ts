export type State = {
    todos: Array<string>;
    todoInputValue: string;
    header: string,
    doneTodos: Array<string>;
    shouldFetchTodos: boolean;
}

export const getInitialState = (): State => {
    return { 
        todos: [],
        todoInputValue: '',
        header: 'Att GÃ¶ra',
        doneTodos: [],
        shouldFetchTodos: true,
     };
}

export const handleNewInputValue = (state: State, v: string): State => {
    state.todoInputValue = v;
    return state;
}

export const addItem = (state: State): State => {
    state.todos.push(state.todoInputValue);
    state.todoInputValue = '';
    return state;
};

export const reverse = (state: State): State => {
    state.todos = state.todos.reverse();
    return state;
}

export const markAsDone = (state: State, done: string): State => {
    state.todos = state.todos.filter((item: string) => item !== done);
    state.doneTodos.push(done);
    return state;
}

export const removeDone = (state: State, done: string): State => {
    state.doneTodos = state.doneTodos.filter((item: string) => item !== done);
    return state
}

export const shouldFetchTodos = (state: State): boolean => {
    return state.shouldFetchTodos;
}

export const receiveTodosResponse = (state: State, response: Array<any>): State => {
    state.shouldFetchTodos = false;
    state.todos = response.map((item: {title: string}) => item.title);
    return state;
}