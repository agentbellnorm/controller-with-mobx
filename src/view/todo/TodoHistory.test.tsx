import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import todoHistoryMaker from './TodoHistory';
import Bucket from '../../plumbing/Bucket';
import * as core from './core';

const shallowRender = (element: React.ReactElement<any, any>) => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();;
} 

describe('TodoHistory', () => {
    it('renders', () => {
        const store = new Bucket<core.State>(core.getInitialState());
        const TodoHistory = todoHistoryMaker(store);
        expect(shallowRender(<TodoHistory />)).toMatchSnapshot();
    });

    it('items in list', () => {
        const store = new Bucket<core.State>({...core.getInitialState(), doneTodos: ['vacuum', 'get food']});
        const TodoHistory = todoHistoryMaker(store);
        expect(shallowRender(<TodoHistory />)).toMatchSnapshot();
    });
});

export {}