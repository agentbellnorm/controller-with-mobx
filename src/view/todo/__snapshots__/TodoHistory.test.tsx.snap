// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoHistory items in list 1`] = `
<React.Fragment>
  <h2>
    Färdiga
  </h2>
  <ul>
    <li
      role="done-item"
    >
      vacuum
      <button
        onClick={[Function]}
      >
        <span
          aria-label="remove"
          role="img"
        >
          ❌
        </span>
      </button>
    </li>
    <li
      role="done-item"
    >
      get food
      <button
        onClick={[Function]}
      >
        <span
          aria-label="remove"
          role="img"
        >
          ❌
        </span>
      </button>
    </li>
  </ul>
</React.Fragment>
`;

exports[`TodoHistory renders 1`] = `
<React.Fragment>
  <h2>
    Färdiga
  </h2>
  <ul />
</React.Fragment>
`;
